File: 300. Longest Increasing Subsequence.cpp
class Solution {
public:
    // dp[i]: length of LIS ending at j
    // O(N2)time, O(N)space, 57.19%, ok
    int lengthOfLIS(vector<int>& nums) {
        if(nums.empty()) return 0;
        int ret = 1;
        vector<int> len(nums.size(), 1); // base case
        for(int i = 1; i < nums.size(); i++){
            for(int j = 0; j < i; j++){
                if(nums[i] > nums[j])
                    len[i] = max(len[i], len[j] + 1);
            }
            ret = max(ret, len[i]);
        }
        return ret;
    }
};

File: 300. Longest Increasing Subsequence.py
class Solution(object):
    
    # dynamic programming: dp[i] = longest length ending in i
    # O(N2) time, O(N) space
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        length = [1 for i in range(len(nums))]
        ret = 1
        for end in xrange(1, len(nums)):
            for start in xrange(end):
                if nums[end] > nums[start]:
                    length[end] = max(length[end], length[start] + 1)
            ret = max(ret, length[end])
        return 0 if len(nums) == 0 else ret
    
    # !!?? dp with binary search: O(NlogN) time, O(N) space
    def lengthOfLIS(self, nums):
       
        